File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeInMemoryVectorStore.kt
package io.github.rajibsarwar.edgeframeworks

class EdgeInMemoryVectorStore : EdgeVectorStore {
    private data class Entry(
        val chunk: EdgeChunk,
        val embedding: EdgeEmbedding
    )

    private val lock = Any()
    private val entries = linkedMapOf<String, Entry>()

    override suspend fun upsert(
        chunks: List<EdgeChunk>,
        embeddings: List<EdgeEmbedding>
    ) {
        if (chunks.size != embeddings.size) {
            throw EdgeVectorException.CountMismatch(
                expected = chunks.size,
                actual = embeddings.size
            )
        }

        synchronized(lock) {
            chunks.zip(embeddings).forEach { (chunk, embedding) ->
                entries[chunk.id] = Entry(
                    chunk = chunk,
                    embedding = embedding
                )
            }
        }
    }

    override suspend fun search(
        query: EdgeEmbedding,
        topK: Int,
        filter: EdgeVectorFilter?
    ): List<EdgeSearchResult> {
        if (topK <= 0) return emptyList()

        val snapshot = synchronized(lock) {
            entries.values
                .filter { entry ->
                    filter?.matches(entry.chunk) ?: true
                }
                .toList()
        }

        return snapshot
            .map { entry ->
                EdgeSearchResult(
                    chunk = entry.chunk,
                    score = EdgeVectorMath.cosineSimilarity(
                        query,
                        entry.embedding
                    )
                )
            }
            .sortedWith(
                compareByDescending<EdgeSearchResult> { it.score }
                    .thenBy { it.chunk.id }
            )
            .take(topK)
    }

    override suspend fun remove(
        filter: EdgeVectorFilter
    ): Int {
        return synchronized(lock) {
            val matchingIds = entries
                .filterValues { entry ->
                    filter.matches(entry.chunk)
                }
                .keys
                .toList()

            matchingIds.forEach(entries::remove)
            matchingIds.size
        }
    }

    override suspend fun removeAll() {
        synchronized(lock) {
            entries.clear()
        }
    }
}

File: ios/Sources/EdgeFrameworks/Runtime/EdgeInMemoryVectorStore.swift
public actor EdgeInMemoryVectorStore: EdgeVectorStore {
    private struct Entry: Sendable {
        let chunk: EdgeChunk
        let embedding: EdgeEmbedding
    }

    private var entries: [String: Entry] = [:]

    public init() {}

    public func upsert(
        chunks: [EdgeChunk],
        embeddings: [EdgeEmbedding]
    ) async throws {
        guard chunks.count == embeddings.count else {
            throw EdgeVectorError.countMismatch(
                expected: chunks.count,
                actual: embeddings.count
            )
        }

        for (chunk, embedding) in zip(chunks, embeddings) {
            entries[chunk.id] = Entry(
                chunk: chunk,
                embedding: embedding
            )
        }
    }

    public func search(
        query: EdgeEmbedding,
        topK: Int,
        filter: EdgeVectorFilter?
    ) async throws -> [EdgeSearchResult] {
        guard topK > 0 else { return [] }

        var results: [EdgeSearchResult] = []
        results.reserveCapacity(entries.count)

        for entry in entries.values {
            if let filter, !filter.matches(entry.chunk) {
                continue
            }
            let score = try EdgeVectorMath.cosineSimilarity(
                query,
                entry.embedding
            )
            results.append(
                EdgeSearchResult(
                    chunk: entry.chunk,
                    score: score
                )
            )
        }

        return Array(
            results
                .sorted { lhs, rhs in
                    if lhs.score == rhs.score {
                        return lhs.chunk.id < rhs.chunk.id
                    }
                    return lhs.score > rhs.score
                }
                .prefix(topK)
        )
    }

    public func remove(
        filter: EdgeVectorFilter
    ) async throws -> Int {
        let matchingIDs = entries
            .filter { filter.matches($0.value.chunk) }
            .map(\.key)

        for id in matchingIDs {
            entries.removeValue(forKey: id)
        }

        return matchingIDs.count
    }

    public func removeAll() async {
        entries.removeAll()
    }
}
